libopenage/gamestate/resource.h

```diff
@@ -6,25 +6,109 @@
 
 namespace openage {
 
+class ResourceBundle;
+
+/**
+ * A resource
+ */
+class Resource {
+public:
+
+	Resource();
+
+	virtual int id() const = 0;
+
+	virtual std::string name() const = 0;
+
+	// TODO add images and icons
+
+};
+
+class ResourceProducer : public Resource {
+public:
+
+	ResourceProducer(int id, std::string name)
+		:
+		_id{id},
+		_name{name} { }
+
+	int id() const override { return _id; }
+
+	std::string name() const override { return _name; }
+
+private:
+
+	int _id;
+	std::string _name;
+};
+
 /**
- * probably becomes part of the nyan game spec in future
+ * All the resources
  */
+class Resources {
+public:
+
+	Resources();
+
+	virtual unsigned int get_count() const = 0;
+
+	virtual const Resource& get_resource(int id) const = 0;
+
+	ResourceBundle create_bundle() const;
+
+};
+
+class ClassicResources : public Resources {
+public:
+
+	ClassicResources()
+		:
+		wood{0, "wood"},
+		food{1, "food"},
+		gold{2, "gold"},
+		stone{3, "stone"},
+		resources{{0, "wood"}, {1, "food"}, {2, "gold"}, {3, "stone"}}{
```

Comment:
why double them?

Reply:
I will keep only the array, but is this model enough?

```c++
class Resources {
public:
	Resources();
	virtual unsigned int get_count() const = 0;
	virtual const Resource& get_resource(int id) const = 0;
	ResourceBundle create_bundle() const; // not virtual
};
```

Reply:
Can't really tell, we only have to design it so that resources can be added or removed at runtime by the API.